File: tigerml/lib/straight_line.ml
type id = string
type binop = Plus | Minus | Times | Div

type stm =
  | CompoundStm of stm * stm
  | AssignStm of id * exp
  | PrintStm of exp list

and exp =
  | IdExp of id
  | NumExp of int
  | OpExp of exp * binop * exp
  | EseqExp of stm * exp

let exampleProg =
  CompoundStm
    ( AssignStm ("a", OpExp (NumExp 5, Plus, NumExp 3)),
      CompoundStm
        ( AssignStm
            ( "b",
              EseqExp
                ( PrintStm [ IdExp "a"; OpExp (IdExp "a", Minus, NumExp 1) ],
                  OpExp (NumExp 10, Times, IdExp "a") ) ),
          PrintStm [ IdExp "b" ] ) )

type envir = (id * int) list

let rec lookup env x =
  match env with
  | [] -> failwith "Variable not found"
  | (y, v) :: rest -> if x = y then v else lookup rest x

let rec evalExp (exp : exp) (env : envir) : envir * int =
  match exp with
  | IdExp id -> (env, lookup env id)
  | NumExp n -> (env, n)
  | OpExp (e1, op, e2) ->
      let env', v1 = evalExp e1 env in
      let env'', v2 = evalExp e2 env' in
      let v =
        match op with
        | Plus -> v1 + v2
        | Minus -> v1 - v2
        | Times -> v1 * v2
        | Div -> v1 / v2
      in
      (env'', v)
  | EseqExp (s, e) ->
      let env', _ = evalStm s env in
      evalExp e env'

and printExpVal (exp : exp) (env : envir) : unit =
  let _, v = evalExp exp env in
  print_int v;
  print_string " "

and evalStm (stm : stm) (env : envir) : envir * int =
  match stm with
  | CompoundStm (s1, s2) ->
      let env', _ = evalStm s1 env in
      evalStm s2 env'
  | AssignStm (id, exp) ->
      let env', v = evalExp exp env in
      ((id, v) :: env', v)
  | PrintStm exps ->
      List.iter (fun e -> printExpVal e env) exps;
      print_newline ();
      (env, 0)

let envToStr (env : envir) : string =
  List.fold_left
    (fun acc (id, v) -> acc ^ id ^ " = " ^ string_of_int v ^ ", ")
    "" env

let interp (prog : stm) : unit =
  let env, v = evalStm prog [] in
  Printf.printf "Program executed successfully\n";
  Printf.printf "Final environment: %s\n" (envToStr env);
  Printf.printf "Final value: %d\n" v

let main () = interp exampleProg

File: tigerrs/src/straight_line.rs
type Id = String;
type Env = Vec<(Id, i32)>;

macro_rules! printExp {
    ($($x:expr),*) => {
        Stm::Print(vec![$($x),*])
    };
}

fn lookup(id: &Id, env: &Env) -> i32 {
    env.iter()
        .rev()
        .find(|(x, _)| x == id)
        .map(|(_, v)| *v)
        .unwrap()
}

fn concat(env: &Env, id: Id, val: i32) -> Env {
    let mut new_env = env.clone();
    new_env.push((id, val));
    new_env
}

enum BinOp {
    Plus,
    Minus,
    Times,
    Div,
}

enum Stm {
    Compound(Box<Stm>, Box<Stm>),
    Assign(Id, Exp),
    Print(Vec<Exp>),
}

enum Exp {
    Id(Id),
    Num(i32),
    Op(Box<Exp>, BinOp, Box<Exp>),
    Eseq(Box<Stm>, Box<Exp>),
}

fn exampl_prog() -> Stm {
    compound(
        assign("a", op(Exp::Num(5), BinOp::Plus, Exp::Num(3))),
        compound(
            assign(
                "b",
                eseq(
                    printExp!(id("a"), op(id("a"), BinOp::Minus, Exp::Num(1))),
                    op(Exp::Num(10), BinOp::Times, id("a")),
                ),
            ),
            Stm::Print(vec![id("b")]),
        ),
    )
}

fn compound(s1: Stm, s2: Stm) -> Stm {
    Stm::Compound(Box::new(s1), Box::new(s2))
}
fn assign(id: &str, exp: Exp) -> Stm {
    Stm::Assign(id.to_string(), exp)
}
fn op(left: Exp, op: BinOp, right: Exp) -> Exp {
    Exp::Op(Box::new(left), op, Box::new(right))
}
fn id(id: &str) -> Exp {
    Exp::Id(id.to_string())
}
fn eseq(stm: Stm, exp: Exp) -> Exp {
    Exp::Eseq(Box::new(stm), Box::new(exp))
}

fn eval_exp(exp: &Exp, env: Env) -> (Env, i32) {
    match exp {
        Exp::Id(id) => {
            let v = lookup(id, &env);
            (env, v)
        }
        Exp::Num(num) => (env, *num),
        Exp::Op(left, op, right) => {
            let (env, l) = eval_exp(left, env);
            let (env, r) = eval_exp(right, env);
            let v = match op {
                BinOp::Plus => l + r,
                BinOp::Minus => l - r,
                BinOp::Times => l * r,
                BinOp::Div => l / r,
            };
            (env, v)
        }
        Exp::Eseq(stm, exp) => {
            let (env, _) = eval_stm(stm, env);
            eval_exp(exp, env)
        }
    }
}

fn eval_stm(stm: &Stm, env: Env) -> (Env, i32) {
    match stm {
        Stm::Compound(s1, s2) => {
            let (env, _) = eval_stm(s1, env);
            eval_stm(s2, env)
        }
        Stm::Assign(id, exp) => {
            let (env, v) = eval_exp(exp, env);
            (concat(&env, id.clone(), v), v)
        }
        Stm::Print(exps) => {
            let (env, v) = exps.iter().fold((env, 0), |(env, _), exp| {
                let (env, v) = eval_exp(exp, env);
                print!("{} ", v);
                (env, 0)
            });
            println!();
            (env, v)
        }
    }
}

fn env_to_string(env: &Env) -> String {
    env.iter()
        .rev()
        .map(|(id, v)| format!("{} = {}", id, v))
        .collect::<Vec<String>>()
        .join(", ")
}

fn interp(stm: Stm) {
    let (env, v) = eval_stm(&stm, vec![]);
    println!("Program executed successfully");
    println!("Final environment: {}", env_to_string(&env));
    println!("Return value: {}", v);
}

pub fn main() {
    interp(exampl_prog());
}
